### ISSUE TITLE
Rule Change: "No uncessary condition" shouldn't happen when checking for existence of a key in an object

### ISSUE THREAD

**Issue by Xavier59** [eslint/eslint#16955]
### What rule do you want to change?

@typescript-eslint/no-unnecessary-condition

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
type Bob = { [x: string]: string };
let bob: Bob = { "hello": "world", "foo": "bar" }

let a = "baz";
if( bob[a] === undefined ) {
   console.log("do a thing");
}
```

### What does the rule currently do for this code?

Warning : Unnecessary conditional, the types have no overlap  @typescript-eslint/no-unnecessary-condition

### What will the rule do after it's changed?

No warning given

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

Alternative solution to avoid generating a warning is the following:

```ts
if(Object.keys(bob).indexOf(a) !== -1){
    console.log("do a thing");
}
```

However, this is not optimized at all. It would be better if we could check if a key is undefined directly.

**Comment by fasttime** [comment]
Hi @Xavier59, the rule you mention is provided by [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint). You may want to report your issue there.
This repo is only for ESLint and its built-in rules, and there is not much we can do to address problems with external plugins, so we'll have to close this issue.
